app/com/linkedin/drelephant/ElephantRunner.java

```diff
@@ -45,136 +48,377 @@
+      for (int i = 0; i < _completedExecutorCount; i++) {
+        _completedJobPool.submit(new CompletedExecutorThread(i + 1, _completedJobQueue));
+      }
+    }
+
+    logger.info("The number of threads analysing running jobs is " + _runningExecutorCount);
+    if (_runningExecutorCount > 0) {
+      _undefinedJobPool = Executors.newFixedThreadPool((int) _runningExecutorCount);
+      for (int i = 0; i < _runningExecutorCount; i++) {
+        _undefinedJobPool.submit(new UndefinedExecutorThread(i + 1, _undefinedJobQueue));
+      }
+    }
+  }
+
+  /**
+   * Fetch all the completed jobs and jobs in <i>undefined</i> state from the resource manager
+   * whose start time lie between <i>from</i> and <i>to</i>.
+   *
+   * Ideally we want to fetch only the completed and running jobs but since the resource manager
+   * api doesn't allow filtering by running job, we will fetch all the jobs in <i>undefined</i>
+   * state and then poll them.
+   *
+   * @param from Lower limit on the start time of the job in millisec
+   * @param to Upper limit on the start time of the job ib millisec
+   */
+  private void fetchApplications(long from, long to) {
+    logger.info("Fetching all the analytic apps which started between " + from + " and " + to);
+
+    List<AnalyticJob> undefinedJobs;
+    List<AnalyticJob> completedJobs;
+    try {
+      _analyticJobGenerator.updateAuthToken(to - _fetchLag);
+
+      undefinedJobs = _analyticJobGenerator.fetchUndefinedAnalyticJobs(from, to);
+      completedJobs = _analyticJobGenerator.fetchCompletedAnalyticJobs(from, to);
+
+      Set<AnalyticJob> commonJobs = Utils.getIntersection(undefinedJobs, completedJobs);
+      if (!commonJobs.isEmpty()) {
+        // Make sure no job belongs to both queues. This may happen when a job completes(succeeded/failed)
+        // after being added to the undefined queue and before being added to the completed queue.
+        undefinedJobs.removeAll(commonJobs);
+      }
+    } catch (Exception e) {
+      logger.error("Error fetching job list. Try again later...", e);
+      return;
+    }
+
+    // There is a lag of job data from AM/NM to JobHistoryServer HDFS, we shouldn't use the current time, since
+    // there might be new jobs arriving after we fetch jobs. We provide one minute delay to address this lag.
+    for (int i = 0; i < completedJobs.size(); i++) {
+      completedJobs.get(i).updateExpiryTime(_fetchLag);
+    }
+
+    logger.info("Completed Job queue size is " + (_completedJobQueue.size() + completedJobs.size()));
+    logger.info("Undefined Job queue size is " + (_undefinedJobQueue.size() + undefinedJobs.size()));
 
+    _completedJobQueue.addAll(completedJobs);
+    _undefinedJobQueue.addAll(undefinedJobs);
+
+    _lastTime = to;
+  }
+
+  private class CompletedExecutorThread implements Runnable {
     private int _threadId;
-    private BlockingQueue<AnalyticJob> _jobQueue;
+    private DelayQueue<AnalyticJob> _completedJobQueue;
```

Comment:
When a job completes, there is a delay/lag in the job data moving to the job history server from the NM. I included a comment in line 242, where the expiry time is updated for a completed job.